[RDY] API: Implement nvim_win_set_buf

HiPhish opened the pull request:
The implementation is extremely crude and certainly missing tons of cleanup, but it works on paper.

janlazo committed:
vim-patch:8.0.1423: error in return not caught by try/catch

Problem:    Error in return not caught by try/catch.
Solution:   Call update_force_abort(). (Yasuhiro Matsomoto, closes vim/vim#2483)
https://github.com/vim/vim/commit/fabaf753e26df5a89a854673d14c15a1fa6b321a

bfredl committed:
TUI: always use safe cursor movement after resize

The old code could lead to a memory error in the following situation:

0. The previous cursor position was row 50 since before, on a grid
larger than 50 rows.
1. grid_resize changes the grid height to 40, and invalidly assumes the
resize moved the physical cursor to row 0
2. Some event used a operation that could move the cursor (such as clear), and
then reset the cursor to the "true" position row 50 (pointless after #8221, but
I forgot to remove it)
3. raw_line/cheap_to_print is invoked, and tries to inspect the grid at
row 50 (memory error)
4. grid_cursor_goto would have been called at this point, and set a
valid cursor position 0-39.

bfredl committed:
TUI: delete "first-row" workaround after resize

This was caused by cursor position being invalid right
after tui_grid_resize, which is now fixed

janlazo committed:
vim-patch:8.0.1807: function to set terminal name is too long

Problem:    Function to set terminal name is too long.
Solution:   Refactor the function.  Fix typo in test.
https://github.com/vim/vim/commit/69e056915c4145b7b64c60963797692a5b293561

bfredl committed:
test: replace wait() with pre-assertion in assert_term_colors

janlazo committed:
vim-patch:8.1.0454: resolve() was not tested with a symlink cycle

Problem:    resolve() was not tested with a symlink cycle.
Solution:   Add a test. (Dominique Pelle, closes vim/vim#3513)
https://github.com/vim/vim/commit/261099070940fb64d4633a766ae3510fffe4171d

janlazo committed:
vim-patch:8.1.0453: MS-Windows: executable() is not reliable

Problem:    MS-Windows: executable() is not reliable.
Solution:   Use $PATHEXT properly. (Yasuhiro Matsumoto, closes vim/vim#3412)
https://github.com/vim/vim/commit/8295666dc2c65e42135b91d5c61e2a140d002333

janlazo committed:
oldtests: win: fix executable() assertions

Windows has "Read and execute" permission via ACL
but nvim and libuv do not support ACL.
Windows does not support the executable bit in chmod-style permissions
but it is safe to assume that if the file exists and is readable,
then it is most likely executable.
This means that win.ini and shell32.dll are "executable"
because they exist, are readable, and are in PATH.

PATHEXT does not affect the executable permission of a file;
it exists to run files on the shell while omitting the file extension.
Assume that PATHEXT is intended for cmd.exe only
because powershell can execute powershell files (ie. *.ps1)
without changing PATHEXT or related cmd.exe environment variable.

In the future, nvim should check the outputs of 'assoc' and 'ftype',
cmd.exe internal commands, or check the registry.
Powershell can be used for ACL if C++/C# API is too difficult to use.

janlazo committed:
vim-patch:8.1.0459: Test_executable fails when there is a dog in the system

Problem:    Test_executable fails when there is a dog in the system.
Solution:   Rename the dog. (Hirohito Higashi)
https://github.com/vim/vim/commit/a05a0d325c7615439f4a42f00682b2ebad43c8d9

janlazo committed:
vim-patch:8.1.0463: "simalt ~x" in .vimrc blocks swap file prompt

Problem:    "simalt ~x" in .vimrc blocks swap file prompt.
Solution:   Flush buffers before prompting. (Yasuhiro Matsumoto,
            closes vim/vim#3518, closes vim/vim#2192)
https://github.com/vim/vim/commit/798184cc67036285a24b38c0bf6668d4d1525548

janlazo committed:
vim-patch:8.1.0466: autocmd test fails

Problem:    Autocmd test fails.
Solution:   Do call inchar() when flushing typeahead.
https://github.com/vim/vim/commit/6a2633b00bb00bcf0d994f08d1c54ace2c221b58

janlazo committed:
vim-patch:8.0.1779: deleting in a block selection causes problems (#9099)

Problem:    Deleting in a block selection causes problems.
Solution:   Check the length of the line before adding bd.textcol and
            bd.textlen. (Christian Brabandt, closes vim/vim#2825)
https://github.com/vim/vim/commit/35e802e713382d7e76232ad344af7dcd577e43de

HiPhish committed:
Implement API method nvim_win_set_buf

This new API method allows setting the buffer of a window based on the
window ID. It complements the existing `nvim_win_get_buf`.

HiPhish committed:
Implement API method nvim_win_set_buf

This new API method allows setting the buffer of a window based on the
window ID. It complements the existing `nvim_win_get_buf`.

HiPhish commented:
The current implementation does change the buffer, but the user needs to switch to another window first to see the changes. I'm certain that just changing the buffer the window points to cannot be all there is to it. The `do_buffer` function is nearly 300 lines long, a good chunk of it is just figuring out what it needs to do based on the arguments, but there is also a lot of cleanup.

I have looked into the implementation of `nvim_set_current_buf`, but its use of `do_buffer` assumes that we want to set the buffer in the current window, not just any arbitrary window, so I cannot use its implementation.

As a side-note, `nvim_set_current_buf` could then be implemented on top of `nvim_win_set_buf` in VimScript as
~~~vim
function! nvim_set_current_buf(buffer)
    call nvim_win_set_buf(nvim_get_current_win(), a:buffer)
endfunction
~~~

bfredl commented:
> I have looked into the implementation of nvim_set_current_buf, but its use of do_buffer assumes that we want to set the buffer in the current window, not just any arbitrary window, so I cannot use its implementation.

For better and worse, a lot of API (and recent vimL functionality) is implemented by changing curwin and/or curbuf, call an existing function that assumes curwin/curbuf, and then restoring the old current values. So I wouldn't be this quick to conclude that `do_buffer` can't be used.

justinmk commented:
> but the user needs to switch to another window first to see the changes.

Most likely that indirectly runs do_buffer, accidentally doing the necessary fixups.

> nvim_set_current_buf, but its use of do_buffer assumes that we want to set the buffer in the current window

Why not set the current window temporarily, then unwind it? That's the usual way. Disabling autocommands etc makes it fast.

HiPhish commented:
> For better and worse, a lot of API (and recent vimL functionality) is implemented by changing curwin and/or curbuf, call an existing function that assumes curwin/curbuf, and then restoring the old current values.

> Why not set the current window temporarily, then unwind it? That's the usual way.

And for a moment there I though I could get away from Vim's ritualistic code by going to the actual source. It's rituals all the way down, isn't it?

HiPhish commented:
> Well, the insides can be ugly, but we hide the ugly behind a nice API. Better to get it right before worrying about elegance.

You are right. My problem with ritualistic code is that it is easy to miss part of the ritual and have things work fine for the most part, but not always. Take for example the updated code I just pushed, it works well, but the error handling is messed up. Looking at the function, it would have been better if I had just written it straight up in VimScript instead.
~~~vim
function! nvim_win_set_buf(win, buf)
    let old_win = nvim_get_current_win();
    call nvim_set_current_win(a:win)
    call nvim_set_current_buf(a:buf)
    call nvim_set_current_win(old_win)
endfunction
~~~
I had hoped that writing this directly in the source would avoid triggering fifty side effects from switching windows, but that would require digging even deeper into Neovim's core code and going down to the bare bytes. At this point I'm not sure if there is even a point in continuing with this anymore.

janlazo committed:
lint

bfredl commented:
> I had hoped that writing this directly in the source would avoid triggering fifty side effects from switching windows, but that would require digging even deeper into Neovim's core code and going down to the bare bytes. At this point I'm not sure if there is even a point in continuing with this anymore.

No need to be so defeatist this quickly. "Ritual" or not, other API functions have managed to avoid filthy side effects without too much raw byte banging. If we look at `nvim_win_set_option` (the arguably most side effect-full operation so far), and dig down the implementation in `api/private/helpers.c`, we find that it uses `switch_win`/`restore_win`, which is purposely built for this save/restore situation. It will probably work better that `goto_tabpage_win` which does implement user visible switching, and indeed would be equivalent to the vimscript visible functions.

HiPhish commented:
@bfredl You are right, I guess I'm just in a bad mood and threw in the towel too early. `switch_win`/`restore_win` seems to do the trick, looking for comments on my implementation.

janlazo committed:
vim-patch:8.1.0468: MS-Windows: filter command with pipe character fails (#9101)

Problem:    MS-Windows: Filter command with pipe character fails. (Johannes
            Riecken)
Solution:   Find the pipe character outside of quotes. (Yasuhiro Matsumoto,
            closes vim/vim#1743, closes vim/vim#3523)
https://github.com/vim/vim/commit/0664089eccec1083dd04ef2255856fb34ce62f15

justinmk commented:
Much better! Just needs a couple tests and address the TODO. Might want to add another sentence or two in the docstring explaining the motivation or an example.

HiPhish commented:
I noticed something weird:
~~~vim
" There are two windows: 1000 and 1001, I am in window 1001 at the moment
call nvim_win_set_option(1000, 'stl', 'Hello world')
" The above changes the status line in the other window, as expected
call nvim_win_set_buf(1000, 2)
" The above changes the buffer, as expected
~~~
Now both windows show the same status line. Switching back to the original buffer restores my `Hello world` status line. It is as if the option is local to the buffer, not the window. Is this correct? The same happens if I stay inside the window I want to change and use `nvim_set_current_buf()`.

> Just needs a couple tests and address the TODO.

Sure thing, I just wanted to get this to work before applying the polish. What should be the API level? 6?

justinmk commented:
>  It is as if the option is local to the buffer, not the window. Is this correct?

Not sure. How does it behave with VimL?

> What should be the API level? 6?

See NVIM_API_LEVEL in CMakeLists.txt

janlazo committed:
vim-patch:8.0.1837: one character cmdline abbreviation not triggered after '<,'>

Problem:    One character cmdline abbreviation not triggered after '<,'>.
Solution:   Skip over the special range. (Christian Brabandt, closes vim/vim#2320)
https://github.com/vim/vim/commit/5e3423d192bfa502c6704f731fa2ec6821f9a2f0

janlazo committed:
lint

janlazo committed:
vim-patch:8.0.1777: cannot cleanup before loading another colorscheme (#9104)

Problem:    Cannot cleanup before loading another colorscheme.
Solution:   Add the ColorSchemePre autocommand event.
https://github.com/vim/vim/commit/60a68362aa73f4a6cb534688978f9dc2b16e60fe

HiPhish commented:
> Not sure. How does it behave with VimL?

That was VimL. Nevermind anyway, I looked into how options are set:
~~~c
// File src/nvim/private/helpers.c
// Omitting irrelevant code
static void set_option_value_for(char *key,
                                 int numval,
                                 char *stringval,
                                 int opt_flags,
                                 int opt_type,
                                 void *from,
                                 Error *err)
{
  switch (opt_type)
  {
    case SREQ_WIN:
      switch_win(&save_curwin, &save_curtab, (win_T *)from, win_find_tabpage((win_T *)from), false) 
      set_option_value_err(key, numval, stringval, opt_flags, err);
      restore_win(save_curwin, save_curtab, true);
      break;
    case SREQ_BUF:
      switch_buffer(&save_curbuf, (buf_T *)from);
      set_option_value_err(key, numval, stringval, opt_flags, err);
      restore_buffer(&save_curbuf);
      break;
  }
}
~~~
There is not really such a thing as a "window-local" option, `nvim_win_set_option(n, 'foo', 'bar')` is really just `nvim_buf_set_option(nvim_win_get_buf(n), 'foo', 'bar')`.

bfredl commented:
Note the docs claim `statusline` is the one and only "global-local window option" (i e it doesn't be have like any other window option), so do expect weird behaviour with it, especially in the API. It seems `nvim_win_set_option(1000, 'stl', 'Hello world')` also sets the buffer local value, so when you visit the buffer in a new window that hasn't seen the buffer before, the local value gets copied. It might be worth trying with some other window option, like `winhl`.

HiPhish commented:
> It might be worth trying with some other window option, like `winhl`.

I tried it and I see the same behaviour: the "window-local" options switches along with the buffer. It makes sense if you look at the code in my previous post: in both cases `set_option_value_err` is called with the same arguments, the only difference is the part of the ritual beforehand and after. For a window we switch to the window in question, thus also switching to its buffer, and for a buffer we switch straight to the buffer in question.

(I hate how Vim as taught me to press CTRL-W to delete the current word, I always end up closing a browser window after having almost finished a post)

bfredl commented:
Ahyes, it likely behaves like `:setl` in both cases, but for a window option it should rather work like plain `:set`, except maybe for `stl` (which will remain an unhappy child in any case)

HiPhish commented:
I have now written a test for the new function and cleaned up linter complaints. Requesting consideration for merging.

bfredl commented:
History looks a bit weird. perhaps rebase the latest commit on master?

bfredl commented on CMakeLists.txt:
```diff
@@ -113,7 +113,7 @@ set(NVIM_VERSION_PATCH 2)
 
 # API level
-set(NVIM_API_LEVEL 5)         # Bump this after any API change.
+set(NVIM_API_LEVEL 6)         # Bump this after any API change.
```
No need to bump this. One can always just use the latest level in a PR; if it needs to be bumped, CI  (or `make functionaltest` locally) will tell you.

HiPhish committed:
Merge branch 'master' of https://github.com/neovim/neovim

HiPhish commented:
Is the history better now? GitHub shows all those other commits from `master` for some reason.

bfredl commented:
I think you merged instead, which doesn't remove these commits. Reset to
https://github.com/neovim/neovim/pull/9100/commits/c55287ad0e6d5c93f80fb09ad0b30867ac895166 again and use `git rebase -i master`. Hopefully it would only show one commit (implement API method), but if the rest remain, just delete those lines.

bfredl commented:
Ref discussion https://github.com/neovim/neovim/pull/9100#issuecomment-429125151 and below.

HiPhish commented:
I followed the instructions, but the history is still messed up. Could it be that I have messed up my `master` when I did `git pull upstream master`? I could do
~~~sh
git reset --hard 384770556ba02433904baf78f499b9bb48010040
~~~
on the `master` branch to get back to the last time before my `git pull` and then do things the right way (whatever the right way might be).

justinmk commented:
@HiPhish I force-pushed to your branch. I used `git merge --squash` to get rid of the unnecessary commits:

    git checkout upstream/master
    git merge --squash refs/pull/upstream/9100
    git push -f https://github.com/HiPhish/neovim HEAD:api-win-set-buf

HiPhish commented:
@justinmk Thank you. What is the Right Way™ of keeping up to data with `master`? This sort of thing keeps happening to me. Here is what I usually do:
~~~sh
git checkout master
git pull upstream master  # upstream is the official repository
git checkout my-feature-branch
git rebase master
~~~

justinmk commented on test/functional/api/window_spec.lua:
```diff
@@ -45,6 +45,16 @@ describe('api/win', function()
+      window('set_buf', windows[2], window('get_buf', windows[1]))
+      eq(window('get_buf', windows[2]), window('get_buf', windows[1]))
+	end)
+  end)
```
indentation

justinmk commented on src/nvim/api/window.c:
```diff
@@ -33,6 +34,44 @@ Buffer nvim_win_get_buf(Window window, Error *err)
+    api_set_error(err,
+                  kErrorTypeException,
+                  "Failed to switch to window %d",
+                  window);
```
should also test all of the failure modes (including the `!win || !buf` case above). In the past we've had hidden bugs until we added test cases for failure modes.

It's easy to test the error cases using `expect_err()`. See https://github.com/neovim/neovim/blob/8fd092f3ff15bf70f84ec0d716c5aaa2c7379fa1/test/functional/api/vim_spec.lua#L28                               
 for example.

HiPhish replied:
I have added tests for invalid window- and buffer ID, but as you said, I have no idea how to make the swap intentionally fail.

HiPhish commented:
*bump* Anything else that needs to be done?

justinmk commented:
Merged after fixing indentation and API level. Thanks @HiPhish ,!
